services/head_tracker.go

```diff
@@ -10,31 +10,26 @@ import (
 	uuid "github.com/satori/go.uuid"
 	"github.com/smartcontractkit/chainlink/logger"
 	"github.com/smartcontractkit/chainlink/store"
+	strpkg "github.com/smartcontractkit/chainlink/store"
 	"github.com/smartcontractkit/chainlink/store/models"
 	"github.com/smartcontractkit/chainlink/store/presenters"
 	"github.com/smartcontractkit/chainlink/utils"
+	"github.com/tevino/abool"
 )
 
-// HeadTrackable represents any object that wishes to respond to ethereum events,
-// after being attached to HeadTracker.
-type HeadTrackable interface {
-	Connect(*models.IndexableBlockNumber) error
-	Disconnect()
-	OnNewHead(*models.BlockHeader)
-}
-
 // HeadTracker holds and stores the latest block number experienced by this particular node
 // in a thread safe manner. Reconstitutes the last block number from the data
 // store on reboot.
 type HeadTracker struct {
-	trackers              map[string]HeadTrackable
+	trackerIDs            map[string]store.HeadTrackable
```

Comment:
I'd propose keeping all HeadTrackable's in a simple array `[]store.HeadTrackable` then where you need random access, either do an indexOf, or just keep a map of offsets by name. This will be faster for your typical use case and should produce simpler code. e.g:

```
trackers []store.HeadTrackable

 func (ht *HeadTracker) onNewHead(head *models.BlockHeader) {
...
 	for _, t := range ht.trackers {
 		t.OnNewHead(head)
 	}
 }
```


hrmm, was looking for an example where the code needed random lookups, but you're already doing `indexOf` in `Detach` which I'm a bit puzzled by. Seems the map is just making matters confusing and shouldn't be needed at all?

Reply:
Sounds good with leaning towards the array. Regarding the map, when removing the HeadTrackables in `Detach`, I would prefer to remove by ID instead of instance. If I only have an array, I would need to do indexOf against pointers which I'm not too keen on. Let me know if there's another way I'm overlooking.

```
Detach(ID string)
v
Detach(trackable HeadTrackable)
```

Reply:
You could just pass the index into the array as the id then instead of a string?

Reply:
After our discussion offline, this has been refactored into the ordered thread safe `attachmentCollection`.